presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,392 @@
+                    newNodeSets.addAll(nodeSets);
+                }
+                newNodeSets.add(newNodeSet.build());
+                nodeSets = newNodeSets;
+            }
+
+            return nodeSets;
+        }
+
+        private static List<JoinNode.EquiJoinClause> flipJoinCriteriaIfNecessary(List<JoinNode.EquiJoinClause> joinCriteria, PlanNode left)
+        {
+            return joinCriteria
+                    .stream()
+                    .map(criterion -> left.getOutputSymbols().contains(criterion.getLeft()) ? criterion : criterion.flip())
+                    .collect(toImmutableList());
+        }
+
+        private static PlanNode getJoinSource(PlanNodeIdAllocator idAllocator, List<PlanNode> nodes, Expression filter, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                if (!(BooleanLiteral.TRUE_LITERAL).equals(filter)) {
+                    return new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return planNode;
+            }
+            return new JoinGraphNode(idAllocator.getNextId(), nodes, filter, outputSymbols);
+        }
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph)
+        {
+            JoinNode join = memo.get(joinGraph);
+            if (join == null) {
+                join = generatePartitions(joinGraph.getSources().size())
+                        .map(partitioning -> createJoinNodeTree(joinGraph, partitioning, idAllocator))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .min(planNodeOrdering)
+                        .orElseThrow(() -> new IllegalStateException("joinOrders cannot be empty"));
+                memo.put(joinGraph, join);
+            }
+            return join;
+        }
+
+        private Optional<JoinNode> createJoinNodeTree(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator)
+        {
+            Optional<JoinNode> joinOptional = createJoinAccordingToPartitioning(joinGraph, partitioning, idAllocator);
+            if (!joinOptional.isPresent()) {
+                return Optional.empty();
+            }
+
+            JoinNode placeholderJoin = joinOptional.get();
+            PlanNode left = placeholderJoin.getLeft();
+            PlanNode right = placeholderJoin.getRight();
+            if (left instanceof JoinGraphNode) {
+                left = chooseJoinOrder((JoinGraphNode) left);
+            }
+            if (right instanceof JoinGraphNode) {
+                right = chooseJoinOrder((JoinGraphNode) right);
+            }
+
+            JoinNode joinNode = (JoinNode) placeholderJoin.replaceChildren(ImmutableList.of(left, right));
+            return Optional.of(chooseJoinNodeProperties(joinNode));
+        }
+
+        private JoinNode chooseJoinNodeProperties(JoinNode joinNode)
```

Comment:
Is it choosing or setting of properties? Right now it seems that only join nodes can have interesting properties but eventually we may wanna associate them with all nodes